### PR TITLE
Ubuntu aws and generic builders

### LINKED ISSUE #4: Builder for ubuntu-generic and ubuntu-aws
**Motivation**

**Feature**

We needd to allow the service to build ubuntu-generic and ubuntu-aws kernels

**Alternatives**

**Additional context**

### PR DESCRIPTION
**Pull request by fntlnz**
Signed-off-by: Lorenzo Fontana <lo@linux.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #4

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new: builder for ubuntu-generic and ubuntu-aws
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,16 +6,65 @@ Status: Under development
...
+#### Vanilla
 ```bash
-echo '{"buildType": "vanilla", "architecture": "x86_64", "kernelversion": "5.5.2", "kernelConfigData": "'"$(zcat /proc/config.gz|base64)"'"}' | 
+echo '{"buildType": "vanilla", "architecture": "x86_64", "kernelrelease": "5.5.2", "kernelConfigData": "'"$(zcat /proc/config.gz|base64)"'"}' | 
     curl  --header "Content-Type: application/json"   -d @-  -v http://127.0.0.1:8093/v1/module
 ```
 
+#### Ubuntu AWS
+```bash
+ echo '{"buildType": "ubuntu-aws", "moduleVersion": "dev", "architecture": "x86_64", "kernelversion": "81", "kernelrelease": "4.15.0-72-generic",  "kernelConfigData": "'"$(cat /boot/config-4.15.0-72-generic |base64)"'"}' |
```
Comment (leodido) [inline README.md]:
```suggestion
 echo '{"buildType": "ubuntu-aws", "moduleVersion": "dev", "architecture": "x86_64", "kernelversion": "81", "kernelrelease": "4.15.0-72-generic",  "kernelConfigData": "'"$(cat /boot/config-4.15.0-72-generic|base64)"'"}' | 
```

Just to be consistent with other examples

File: README.md [hunk 2]
```diff
@@ -6,16 +6,65 @@ Status: Under development
...
 ```bash
-echo '{"buildType": "vanilla", "architecture": "x86_64", "kernelversion": "5.5.2", "kernelConfigData": "'"$(zcat /proc/config.gz|base64)"'"}' | 
+echo '{"buildType": "vanilla", "architecture": "x86_64", "kernelrelease": "5.5.2", "kernelConfigData": "'"$(zcat /proc/config.gz|base64)"'"}' | 
     curl  --header "Content-Type: application/json"   -d @-  -v http://127.0.0.1:8093/v1/module
 ```
 
+#### Ubuntu AWS
+```bash
+ echo '{"buildType": "ubuntu-aws", "moduleVersion": "dev", "architecture": "x86_64", "kernelversion": "81", "kernelrelease": "4.15.0-72-generic",  "kernelConfigData": "'"$(cat /boot/config-4.15.0-72-generic |base64)"'"}' | 
+    curl  --header "Content-Type: application/json"   -d @-  -v http://bdb769cd.ngrok.io/v1/module
```
Comment (leodido) [inline README.md]:
```suggestion
    curl  --header "Content-Type: application/json" -d @-  -v http://bdb769cd.ngrok.io/v1/module
```

File: README.md [hunk 3]
```diff
@@ -6,16 +6,65 @@ Status: Under development
...
+```
+
+#### Ubuntu Generic
+
+```bash
+ echo '{"buildType": "ubuntu-aws", "moduleVersion": "dev", "architecture": "x86_64", "kernelversion": "59", "kernelrelease": "4.15.0-1057-aws",  "kernelConfigData": "'"$(cat /boot/config-4.15.0-1057-aws|base64)"'"}' | 
+    curl  --header "Content-Type: application/json"   -d @-  -v http://bdb769cd.ngrok.io/v1/module
```
Comment (leodido) [inline README.md]:
```suggestion
    curl --header "Content-Type: application/json" -d @- -v http://bdb769cd.ngrok.io/v1/module
```

File: README.md [hunk 4]
```diff
@@ -6,16 +6,65 @@ Status: Under development
...
+```
+
+
 ### Retrieve a built module
 
+When you request the build, the response will give you the final destination.
+In general this is the pattern:
+
+```
+/v1/module/<buildtype>/<architecture>/<module-version>/<kernel-release>/<kernel-version>/<kernel-config-sha256sum>
```
Comment (leodido) [inline README.md]:
```suggestion
/v1/module/<buildtype>/<architecture>/<module-version>/<kernel-release>/<kernel-version>/<kernel-config-sha256sum>/falco.ko
```

File: README.md [hunk 5]
```diff
@@ -35,3 +84,6 @@ curl -O  -v http://127.0.0.1:8093/v1/module/vanilla/x86_64/dev/5.5.2-arch1-1/518
 ## Interactions Diagram
 
 ![Interaction Diagram](docs/img/interactions.png)
+
+https://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux-aws/linux-headers-4.15.0-1057-aws_4.15.0-1057.59_amd64.deb
```
Comment (leodido) [inline README.md]: Why links to linux headers in the REAMDE?

Comment (leodido) [inline README.md]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -35,3 +84,6 @@ curl -O  -v http://127.0.0.1:8093/v1/module/vanilla/x86_64/dev/5.5.2-arch1-1/518
...
+https://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux-aws/linux-headers-4.15.0-1057-aws_4.15.0.59_amd64.deb
```
Comment (leodido) [inline README.md]:
```suggestion
```

File: pkg/filesystem/modulestorage.go [hunk 7]
```diff
@@ -47,14 +47,14 @@ func (ms *ModuleStorage) openModule(name string) (io.ReadCloser, error) {
...
 }
 func moduleFilenameFromModuleRetrieveRequest(b types.ModuleRetrieveRequest) string {
-	return moduleFilenameFromParams(string(b.BuildType), b.Architecture, b.KernelVersion, b.ModuleVersion, b.ConfigSHA256)
+	return moduleFilenameFromParams(string(b.BuildType), b.Architecture, b.KernelRelease, b.KernelVersion, b.ModuleVersion, b.ConfigSHA256)
 }
 
-func moduleFilenameFromParams(buildType, architecture, kernelVersion, moduleVersion, sha256 string) string {
-	return path.Clean(path.Base(fmt.Sprintf("falco-%s-%s-%s-%s-%s.ko", buildType, architecture, kernelVersion, moduleVersion, sha256)))
+func moduleFilenameFromParams(buildType, architecture, kernelRelease, kernelVersion, moduleVersion, sha256 string) string {
+	return path.Clean(path.Base(fmt.Sprintf("falco-%s-%s--%s-%s-%s-%s.ko", buildType, architecture, kernelRelease, kernelVersion, moduleVersion, sha256)))
```
Comment (leodido) [inline pkg/filesystem/modulestorage.go]: Just to be sure: is the double dash `--` wanted here?

If I remember correctly, yes. But, as said, let's just double-check together!

File: pkg/modulebuilder/buildtype/buildtype.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package buildtype
+
+import "github.com/asaskevich/govalidator"
```
Comment (leodido) [inline pkg/modulebuilder/buildtype/buildtype.go]: When you told me I thought you were talking about this https://github.com/go-playground/validator

File: pkg/filesystem/modulestorage.go (hunk 7 above)
Comment (fntlnz) [reply pkg/filesystem/modulestorage.go]: No it's not, I'm removing it

File: pkg/modulebuilder/buildtype/buildtype.go (hunk 8 above)
Comment (fntlnz) [reply pkg/modulebuilder/buildtype/buildtype.go]: Yeah, used the wrong one! Feel free to change it in a new PR

File: pkg/modulebuilder/builder/ubuntu.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	buf := bytes.NewBuffer(nil)
+	err = parsed.Execute(buf, td)
+	if err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func fetchUbuntuGenericKernelURL(kr kernelrelease.KernelRelease, kernelVersion string) []string {
+	panic("not implemented yet")
```
Comment (fntlnz) [inline pkg/modulebuilder/builder/ubuntu.go]: For myself: remember to do this before merge

**Comment by fntlnz** [comment]
Reopening to let poiana do its job

**Pull request merged**